pkg/iter/entry_iterator.go

```diff
@@ -374,35 +405,55 @@ func (i *entrySortIterator) init() {
 		}
 		util.LogError("closing iterator", it.Close)
 	}
-	heap.Init(i.heap)
+	i.is = tmp
+	sort.Slice(i.is, i.less)
+}
 
-	// We can now clear the list of input iterators to merge, given they have all
-	// been processed and the non empty ones have been pushed to the heap
-	i.is = nil
+func (i *entrySortIterator) fix() {
+	t1 := i.is[0].Entry().Timestamp
+	l1 := i.is[0].Labels()
+
+	// shortcut
+	if len(i.is) <= 1 || i.lessThan(t1, l1, 1) {
+		return
+	}
+
+	// First element is out of place. So we reposition it.
+	index := sort.Search(len(i.is), func(in int) bool { return i.lessThan(t1, l1, in) })
+
+	head := i.is[0]
```

Comment:
head could be defined at the top and reused everywhere.

Reply:
Nice that gave the speed up to 25%.